src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -24,26 +24,79 @@ def self.compile(compile_options)
+        # CODEREVIEW: where should these requires be, and what should happen if they are not found?
+        require 'bolt'
+        require 'bolt/target'
+        require 'bolt/apply_inventory'
+        require 'bolt/apply_target'
+        require 'bolt/apply_result'
+        require 'addressable'
+
+        Puppet[:rich_data] = true
         Puppet[:node_name_value] = compile_options['certname']
 
+        # TODO: again, error handling/warning here when this is not supplied
+        boltlib_path = Array(compile_options.dig('options', 'boltlib') || '')
         # Use the existing environment with the requested name
         Puppet::Pal.in_environment(compile_options['environment'],
+                                   pre_modulepath: boltlib_path,
                                    envpath: Puppet[:environmentpath],
-                                   facts: compile_options['facts']['values'],
-                                   variables: compile_options['variables']['values']) do |pal|
-          Puppet.lookup(:pal_current_node).trusted_data = compile_options['trusted_facts']['values']
-
-          # This compiler has been configured with a node containing
-          # the requested environment, facts, and variables, and is used
-          # to compile a catalog in that context from the supplied AST.
-          pal.with_catalog_compiler do |compiler|
-            # We have to parse the AST inside the compiler block, because it
-            # initializes the necessary type loaders for us.
-            ast = Puppet::Pops::Serialization::FromDataConverter.convert(code)
-
-            compiler.evaluate(ast)
-            compiler.compile_additions
-            compiler.catalog_data_hash
+                                   facts: compile_options['facts']['values']) do |pal|
+          # No conn info for PE targets can just use empty hash if 
+          # https://github.com/puppetlabs/bolt/pull/1770 is accepted
+          fake_config = {
+            'transport' => 'redacted',
+            'transports' => {
+              'redacted' => 'redacted'
+            }
+          }
+          bolt_inv = Bolt::ApplyInventory.new(fake_config)
+          Puppet.override(bolt_inventory: bolt_inv) do
+            Puppet.lookup(:pal_current_node).trusted_data = compile_options['trusted_facts']['values']
+            #TODO: get rid of this debug logging...
+            Puppet.warning("Compile Options")
+            Puppet.warning("#{compile_options}")
+            # This compiler has been configured with a node containing
+            # the requested environment, facts, and variables, and is used
+            # to compile a catalog in that context from the supplied AST.
+            pal.with_catalog_compiler do |compiler|
+              #CODEREVIEW: Do we need to make these configurable in the request?
+              Puppet[:strict] = :warning
+              Puppet[:strict_variables] = false
```

Comment:
These are the defaults in Puppet, but the opposite of PAL, right? Should the PAL defaults just change?

Reply:
I cant find these set anywhere except https://github.com/puppetlabs/puppet/blob/bf06b6258b10c3669e32f670fcd4052ce02e1953/lib/puppet/parser/catalog_compiler.rb#L15-L16 in Puppet.

Reply:
That catalog compiler is the compiler used by PAL (eg, its setting defaults for PAL there, a normal Catalog compilation goes through a different entrypoint).

The default value for strict is here: https://github.com/puppetlabs/puppet/blob/master/lib/puppet/defaults.rb#L197
and strict_variables here: https://github.com/puppetlabs/puppet/blob/master/lib/puppet/defaults.rb#L2131